icu4c/source/common/ucnvmbcs.cpp

```diff
@@ -3144,13 +3144,14 @@ ucnv_MBCSGetNextUChar(UConverterToUnicodeArgs *pArgs,
     }
 
     if(c<0) {
-        if(U_SUCCESS(*pErrorCode) && source==sourceLimit && lastSource<source) {
+        if(U_SUCCESS(*pErrorCode) && source==sourceLimit && lastSource<source &&
+           size_t(source-lastSource) <= sizeof(cnv->toUBytes)) {
             /* incomplete character byte sequence */
             uint8_t *bytes=cnv->toUBytes;
             cnv->toULength = static_cast<int8_t>(source - lastSource);
             do {
                 *bytes++=*lastSource++;
-            } while(lastSource<source);
+            } while(lastSource<source && bytes <= &(cnv->toUBytes[sizeof(cnv->toUBytes)-1]));
```

Comment:
FYI only -- pretty sure that this additional condition is guaranteed true

Reply:
removed

Reply:
Actually, I have to add them back to make my build work. Here are the errors I got if I remove them in my local build

```make[1]: Entering directory '/usr/local/google/home/ftang/icu/icu/icu4c/source/common'
   g++	 ...  ucnvmbcs.cpp
ucnvmbcs.cpp: In function ‘UChar32 ucnv_MBCSGetNextUChar(UConverterToUnicodeArgs*, UErrorCode*)’:
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
In file included from ucnvmbcs.cpp:55:
ucnv_bld.h:190:13: note: at offset 16 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 16 into destination object ‘UConverter::toUBytes’ of size 7
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
ucnv_bld.h:190:13: note: at offset 17 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: at offset [1, 7] into destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 17 into destination object ‘UConverter::toUBytes’ of size 7
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
ucnv_bld.h:190:13: note: at offset 18 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: at offset [2, 7] into destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 18 into destination object ‘UConverter::toUBytes’ of size 7
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
ucnv_bld.h:190:13: note: at offset 19 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: at offset [3, 7] into destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 19 into destination object ‘UConverter::toUBytes’ of size 7
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
ucnv_bld.h:190:13: note: at offset 20 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: at offset [4, 7] into destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 20 into destination object ‘UConverter::toUBytes’ of size 7
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
ucnv_bld.h:190:13: note: at offset 21 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: at offset [5, 7] into destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 21 into destination object ‘UConverter::toUBytes’ of size 7
ucnvmbcs.cpp:3153:25: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 3153 |                 *bytes++=*lastSource++;
      |                 ~~~~~~~~^~~~~~~~~~~~~~
ucnv_bld.h:190:13: note: at offset 22 into destination object ‘UConverter::toUBytes’ of size 7
  190 |     uint8_t toUBytes[UCNV_MAX_CHAR_LEN-1];/* more "toU status"; keeps the bytes of the current character */
      |             ^~~~~~~~
ucnv_bld.h:190:13: note: at offset [6, 7] into destination object ‘UConverter::toUBytes’ of size 7
ucnv_bld.h:190:13: note: at offset 22 into destination object ‘UConverter::toUBytes’ of size 7
cc1plus: all warnings being treated as errors
*** Failed compilation command follows: ----------------------------------------------------------
g++ -D_REENTRANT -DU_HAVE_ELF_H=1 -DU_HAVE_TZNAME=0 -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=0 -I. -DDEFAULT_ICU_PLUGINS="/usr/local/lib/icu"  -DU_ALL_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DU_COMMON_IMPLEMENTATION -O3 -Werror -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long -std=c++17 -c -DPIC -fPIC -o ucnvmbcs.o ucnvmbcs.cpp
--- ( rebuild with "make VERBOSE=1 all" to show all param
```

Reply:
toUBytes[] collects bytes of an incomplete byte sequence for encoding one character. With size 7 for some incomplete byte sequence, the next longer complete byte sequence would be 8 bytes long. I don't think we support any conversion where a single-character byte sequence is longer than 8 bytes.

Is it possible to figure out which converter it is using, and what the input byte sequence is?

If we have to add these checks now, then please create a follow-up ticket to investigate, and add TODO comments next to the new checks.